File: infrastructures/ddd/python/src/infrastructure/persistence/write/example_write_entity.py
"""
Example Write Entity demonstrating command side of CQRS.
This class:
- Lives in /infrastructure/persistence/write/
- Has WriteEntity suffix
- Used for business logic execution (commands)
- Contains all fields needed for business operations
- CAN have database/ORM attributes (e.g., MongoDB decorators, SQLAlchemy columns)
- Mapped from/to BMO by repository using mapper
"""

from datetime import datetime
from typing import Optional
from infrastructure.persistence.write.infra import BaseWriteEntity


class ExampleWriteEntity(BaseWriteEntity):
    """
    Write entity for example domain object.
    
    In a real application, this would be decorated with ORM attributes:
    - MongoDB: @dataclass with field metadata
    - SQLAlchemy: Column definitions
    - Django ORM: Model field definitions
    
    Example with MongoDB (pymongo/motor):
    ```python
    from dataclasses import dataclass, field
    
    @dataclass
    class ExampleWriteEntity(BaseWriteEntity):
        name: str = field(metadata={"bson_field": "name"})
        description: str = field(metadata={"bson_field": "description"})
        owner_id: str = field(metadata={"bson_field": "owner_id"})
        is_active: bool = field(metadata={"bson_field": "is_active"})
    ```
    
    Example with SQLAlchemy:
    ```python
    from sqlalchemy import Column, String, Boolean
    
    class ExampleWriteEntity(BaseWriteEntity, Base):
        __tablename__ = "examples"
        
        name = Column(String(100), nullable=False)
        description = Column(String(500))
        owner_id = Column(String(36), nullable=False)
        is_active = Column(Boolean, default=True)
    ```
    """
    
    def __init__(self):
        """Initialize a new example write entity."""
        super().__init__()
        self.name: str = ""
        self.description: str = ""
        self.owner_id: str = ""
        self.is_active: bool = True

File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/infrastructure/persistence/write/ExampleWriteEntity.java
package com.oddly.ddd.infrastructure.persistence.write;

import com.oddly.ddd.infrastructure.persistence.write.infra.BaseWriteEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;

/**
 * Write entity for Example (command side in CQRS).
 * Write entities:
 * - Live in /infrastructure/persistence/write/
 * - Have WriteEntity suffix
 * - Used for commands (create, update, delete)
 * - Contain all fields needed for business operations
 * - Have ORM/database annotations
 */
@Entity
@Table(name = "examples")
public class ExampleWriteEntity extends BaseWriteEntity {
    @Column(name = "name", nullable = false, length = 100)
    private String m_name;

    @Column(name = "description", length = 500)
    private String m_description;

    @Column(name = "category", length = 50)
    private String m_category;

    @Column(name = "is_active", nullable = false)
    private boolean m_isActive;

    @Column(name = "owner_id", nullable = false)
    private String m_ownerId;

    public ExampleWriteEntity() {
        super();
        this.m_isActive = true;
    }

    public String getName() {
        return m_name;
    }

    public void setName(String p_name) {
        this.m_name = p_name;
    }

    public String getDescription() {
        return m_description;
    }

    public void setDescription(String p_description) {
        this.m_description = p_description;
    }

    public String getCategory() {
        return m_category;
    }

    public void setCategory(String p_category) {
        this.m_category = p_category;
    }

    public boolean isActive() {
        return m_isActive;
    }

    public void setActive(boolean p_active) {
        m_isActive = p_active;
    }

    public String getOwnerId() {
        return m_ownerId;
    }

    public void setOwnerId(String p_ownerId) {
        this.m_ownerId = p_ownerId;
    }
}
